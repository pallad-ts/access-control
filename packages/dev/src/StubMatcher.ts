import {AccessQuery} from "@pallad/access-control";

const deepEqual = require('deep-equal');

const ALWAYS_TRUE = () => true;

export class StubMatcher {
	private principalList = new Set<StubMatcher.Matcher<unknown>>();
	private actionList = new Set<StubMatcher.Matcher<string>>();
	private subjectList = new Set<StubMatcher.Matcher<unknown>>();

	anyAccessQuery() {
		return this.anyPrincipal()
			.anyAction()
			.anySubject()
	}

	anyPrincipal(): this {
		this.principalList.add(ALWAYS_TRUE);
		return this;
	}

	principal(...principal: Array<unknown | ((principal: unknown) => boolean)>): this {
		for (const x of principal) {
			this.principalList.add(x);
		}
		return this;
	}

	anyAction(): this {
		this.actionList.add(ALWAYS_TRUE);
		return this;
	}

	action(...action: Array<string | ((action: string) => boolean)>): this {
		for (const x of action) {
			this.actionList.add(x);
		}
		return this;
	}

	anySubject(): this {
		this.subjectList.add(ALWAYS_TRUE);
		return this;
	}

	subject(...subject: Array<unknown | ((subject: unknown) => boolean)>): this {
		for (const x of subject) {
			this.subjectList.add(x);
		}
		return this;
	}

	matchAccessQuery(query: AccessQuery): boolean {
		return this.matchPrincipal(query.principal) && this.matchAction(query.action) && this.matchSubject(query.subject);
	}

	private matchPrincipal(principal: unknown) {
		this.assertIsConfigured(this.principalList, 'No principal matcher defined. Define list of principals or use `anyParticipant` to match any participant');
		return !!Array.from(this.principalList).find(x => {
			if (x instanceof Function) {
				return x(principal);
			}

			return x === principal || deepEqual(x, principal);
		})
	}

	private assertIsConfigured(set: Set<any>, errorMessage: string) {
		if (set.size === 0) {
			throw new Error(errorMessage);
		}
	}

	private matchAction(action: string) {
		this.assertIsConfigured(this.actionList, 'No action matcher defined. Define list of actions or use `anyAction` to match any action');
		return !!Array.from(this.actionList).find(x => {
			if (x instanceof Function) {
				return x(action);
			}

			return x === action;
		})
	}

	private matchSubject(subject: unknown) {
		this.assertIsConfigured(this.subjectList, 'No subject matcher defined. Define list of subjects or use `anySubject` to match any subject');
		return !!Array.from(this.subjectList).find(x => {
			if (x instanceof Function) {
				return x(subject);
			}

			return x === subject || deepEqual(x, subject);
		})
	}
}

export namespace StubMatcher {
	export type Matcher<T> = T | ((arg: T) => boolean);
}
