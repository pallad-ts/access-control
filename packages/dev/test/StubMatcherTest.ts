import {StubMatcher} from "@src/StubMatcher";
import {ACCESS_QUERY, ACTION, ACTION_2, PRINCIPAL, PRINCIPAL_2, SUBJECT, SUBJECT_2} from "./fixtures";

describe('StubMatcher', () => {

	let matcher: StubMatcher;
	beforeEach(() => {
		matcher = new StubMatcher();
	});

	describe('matching access query', () => {
		it('it is impossible to match access query if none of access query parts matcher is defined', () => {
			expect(() => {
				matcher.matchAccessQuery({
					principal: PRINCIPAL,
					subject: SUBJECT,
					action: ACTION
				})
			})
				.toThrowErrorMatchingSnapshot();
		});
	});

	it('anyAccessQuery effectively allow everything', () => {
		matcher.anyAccessQuery();

		expect(matcher.matchAccessQuery({
			principal: Math.random(),
			action: String(Math.random()),
			subject: Math.random()
		}))
			.toBe(true);
	});

	describe('defining principal', () => {
		beforeEach(() => {
			matcher.subject(SUBJECT)
				.action(ACTION);
		});

		it('fails if principal matcher is not defined', () => {
			expect(() => {
				matcher.matchAccessQuery(ACCESS_QUERY);
			})
				.toThrowErrorMatchingSnapshot();
		});

		it('anyPrincipal cause to match any principal', () => {
			matcher.anyPrincipal();

			expect(matcher.matchAccessQuery({
				action: ACTION,
				subject: SUBJECT,
				principal: Math.random()
			}))
				.toBe(true);
		});

		it('principal matching by deep equality comparison', () => {
			matcher.principal({
				some: 'principal'
			})

			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: ACTION,
				principal: {some: 'principal'}
			}))
				.toBe(true);

			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: ACTION,
				principal: {some: 'principal', withUnnecessary: 'property'}
			}))
				.toBe(false);
		});

		it('multiple principals', () => {
			matcher.principal(PRINCIPAL, PRINCIPAL_2)

			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: ACTION,
				principal: PRINCIPAL
			}))
				.toBe(true);
			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: ACTION,
				principal: PRINCIPAL_2
			}))
				.toBe(true);

			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: ACTION,
				principal: {}
			}))
				.toBe(false);
		});

		it('principal matching by predicate', () => {
			matcher.principal((x: any) => x.some === 'principal')

			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: ACTION,
				principal: {some: 'principal'}
			}))
				.toBe(true);

			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: ACTION,
				principal: {some: 'principal', withUnnecessary: 'property'}
			}))
				.toBe(true);
		})
	});

	describe('defining action', () => {
		beforeEach(() => {
			matcher.subject(SUBJECT)
				.principal(PRINCIPAL);
		});

		it('fails if action matcher is not defined', () => {
			expect(() => {
				matcher.matchAccessQuery(ACCESS_QUERY);
			})
				.toThrowErrorMatchingSnapshot();
		});

		it('anyAction cause to match any action', () => {
			matcher.anyAction();

			expect(matcher.matchAccessQuery({
				action: String(Math.random()),
				subject: SUBJECT,
				principal: PRINCIPAL,
			}))
				.toBe(true);
		});

		it('action matching by strict equality comparison', () => {
			matcher.action('some-action')

			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: 'some-action',
				principal: PRINCIPAL
			}))
				.toBe(true);

			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: 'different-action',
				principal: PRINCIPAL
			}))
				.toBe(false);
		});

		it('multiple actions', () => {
			matcher.action(ACTION, ACTION_2)

			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: ACTION,
				principal: PRINCIPAL
			}))
				.toBe(true);

			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: ACTION_2,
				principal: PRINCIPAL
			}))
				.toBe(true);

			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: 'random-action',
				principal: PRINCIPAL
			}))
				.toBe(false);
		});

		it('action matching by predicate', () => {
			matcher.action(x => x.startsWith('foo_'))

			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: 'foo_zee',
				principal: PRINCIPAL
			}))
				.toBe(true);

			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: 'foo_bar',
				principal: PRINCIPAL
			}))
				.toBe(true);

			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: 'bar_foo',
				principal: PRINCIPAL
			}))
				.toBe(false);
		});
	});

	describe('defining subject', () => {
		beforeEach(() => {
			matcher.action(ACTION)
				.principal(PRINCIPAL);
		});

		it('fails if subject matcher is not defined', () => {
			expect(() => {
				matcher.matchAccessQuery(ACCESS_QUERY);
			})
				.toThrowErrorMatchingSnapshot();
		});

		it('anySubject cause to match any subject', () => {
			matcher.anySubject();

			expect(matcher.matchAccessQuery({
				action: ACTION,
				subject: Math.random(),
				principal: PRINCIPAL,
			}))
				.toBe(true);
		});

		it('subject matching by deep equality comparison', () => {
			matcher.subject({some: 'subject'})

			expect(matcher.matchAccessQuery({
				subject: {some: 'subject'},
				action: ACTION,
				principal: PRINCIPAL
			}))
				.toBe(true);

			expect(matcher.matchAccessQuery({
				subject: {some: 'subject', unnecessaryProperty: 'value'},
				action: ACTION,
				principal: PRINCIPAL
			}))
				.toBe(false);
		});

		it('multiple subjects', () => {
			matcher.subject(SUBJECT, SUBJECT_2)

			expect(matcher.matchAccessQuery({
				subject: SUBJECT,
				action: ACTION,
				principal: PRINCIPAL
			}))
				.toBe(true);

			expect(matcher.matchAccessQuery({
				subject: SUBJECT_2,
				action: ACTION,
				principal: PRINCIPAL
			}))
				.toBe(true);

			expect(matcher.matchAccessQuery({
				subject: {},
				action: ACTION,
				principal: PRINCIPAL
			}))
				.toBe(false);
		});

		it('subject matching by predicate', () => {
			matcher.subject((x: any) => x.some === 'subject')
			expect(matcher.matchAccessQuery({
				subject: {some: 'subject'},
				action: ACTION,
				principal: PRINCIPAL
			}))
				.toBe(true);

			expect(matcher.matchAccessQuery({
				subject: {some: 'subject', unnecessaryProperty: 'value'},
				action: ACTION,
				principal: PRINCIPAL
			}))
				.toBe(true);
		});
	});
});
