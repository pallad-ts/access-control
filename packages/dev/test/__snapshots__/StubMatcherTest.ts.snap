// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StubMatcher defining action fails if action matcher is not defined 1`] = `"No action matcher defined. Define list of actions or use \`anyAction\` to match any action"`;

exports[`StubMatcher defining principal fails if principal matcher is not defined 1`] = `"No principal matcher defined. Define list of principals or use \`anyParticipant\` to match any participant"`;

exports[`StubMatcher defining subject fails if subject matcher is not defined 1`] = `"No subject matcher defined. Define list of subjects or use \`anySubject\` to match any subject"`;

exports[`StubMatcher matching access query it is impossible to match access query if none of access query parts matcher is defined 1`] = `"No principal matcher defined. Define list of principals or use \`anyParticipant\` to match any participant"`;
